src/NUnitEngine/nunit.engine/Services/TestSelectionParser.cs

```diff
@@ -0,0 +1,275 @@
+            }
+
+            if (terms.Count == 1)
+                return terms[0];
+
+            var sb = new StringBuilder("<or>");
+
+            foreach (string term in terms)
+                sb.Append(term);
+
+            sb.Append("</or>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single element or an and expression and return the xml
+        /// </summary>
+        public string ParseFilterTerm()
+        {
+            var elements = new List<string>();
+            elements.Add(ParseFilterElement());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != AND_OP1 && op != AND_OP2)
+                    break;
+
+                NextToken();
+                elements.Add(ParseFilterElement());
+            }
+
+            if (elements.Count == 1)
+                return elements[0];
+
+            var sb = new StringBuilder("<and>");
+
+            foreach (string element in elements)
+                sb.Append(element);
+
+            sb.Append("</and>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single filter element such as a category expression
+        /// and return the xml representation of the filter.
+        /// </summary>
+        public string ParseFilterElement()
+        {
+            Token nextOp = Peek();
+            if (nextOp == LPAREN || nextOp == NOT_OP)
+                return ParseExpressionInParentheses();
+
+            string lhs = GetWord();
+
+            switch (lhs)
+            {
+                case "id":
+                    string op = GetOperator("=", "==");
+                    string values = GetIdList();
+                    return string.Format("<id>{0}</id>", values);
+
+                case "cat":
+                case "category":
+                case "method":
+                case "class":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = GetWordList();
+                    string tag = lhs == "category" ? "cat" : lhs;
+                    switch (op)
+                    {
+                        case "=":
+                        case "==":
+                            return string.Format("<{0}>{1}</{0}>", tag, values);
+                        case "!=":
+                            return string.Format("<not><{0}>{1}</{0}></not>", tag, values);
+                        default:
+                            return string.Format("<{0} op='{1}'>{2}</{0}>", tag, op, values);
+                    }
+
+                case "name":
+                case "test":
+                case "fullname":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = Peek().Kind == TokenKind.String
+                        ? GetString()
+                        : GetWord();
+                    tag = lhs == "fullname" ? "test" : lhs;
```

Comment:
test is a synonym for fullname and cat is for category Is that what you mean?

The other things are not synonyms, but do different things. Of course, one could argue about whether we need them all.

I'm happy to drop category. The word test is already in use as an option, so I included it to make it clear that the syntax test == "My.Test.Method" is equivalent to --test=MyTest.Method. OTOH, I am bothered that our historical name 'test' is so vague. Hence I added fullname. However, it's the full name of the test, not of the method. We should definitely review what we support as the left hand side of a comparison. It's a key.